### PR TITLE
Assign latest virtual timestamp for added node to enable LRU manner of RegexCache

### LINKED ISSUE #105748: Regex internal static cache: new regex instances are treated as 'oldest' ones and removed after cache is full (LRU may not work)
### Description

When I learn from .net internal `RegexCache`, I found possible issue: not sure whether LRU of this cache works properly ? 

When size hits max cache size, newly requested regex instance will only live before next new regex comes, new one is treated as 'oldest' one. Eventually when regex cache is full, basically future waves of new regex will not be cached and need to re-compile when re-request, real old data always occupied cache space. Maybe this may impact expected effect of this cache and performance of related static methods of `Regex` ? 

### Reproduction Steps

```csharp
        static void Main(string[] args)
        {
            TestRegexCache();
        }

        private static void TestRegexCache()
        {
            for (int round = 0; round < 2; round++)
            {
                for (int i = 1; i <= Regex.CacheSize; i++)
                {
                    string pattern = i.ToString();

                    _ = Regex.IsMatch(pattern, pattern);
                }
            }

            string addedPattern1 = "newlyAdded1";
            string addedPattern2 = "newlyAdded2";

            _ = Regex.IsMatch(addedPattern1, addedPattern1);
            Regex addedRegex1 = GetInnerRegexList().Single(r => r.ToString() == addedPattern1);
            _ = Regex.IsMatch(addedPattern2, addedPattern2);
            Regex addedRegex2 = GetInnerRegexList().Single(r => r.ToString() == addedPattern2);

            while (true)
            {
                _ = Regex.IsMatch(addedPattern1, addedPattern1);

                Console.WriteLine($"When add {nameof(addedPattern1)}:");

                Console.WriteLine($"{nameof(addedRegex2)} still in cache list ? - " + GetInnerRegexList().Any(r => r.ToString() == addedPattern2));
                Regex tmp = GetInnerRegexList().Single(r => r.ToString() == addedPattern1);
                Console.WriteLine("Regex instance 1 is reused ? - " + ReferenceEquals(tmp, addedRegex1));
                addedRegex1 = tmp;

                _ = Regex.IsMatch(addedPattern2, addedPattern2);

                Console.WriteLine($"When add {nameof(addedPattern2)}:");

                Console.WriteLine($"{nameof(addedRegex1)} still in cache list ? - " + GetInnerRegexList().Any(r => r.ToString() == addedPattern1));
                tmp = GetInnerRegexList().Single(r => r.ToString() == addedPattern2);
                Console.WriteLine("Regex instance 2 is reused ? - " + ReferenceEquals(tmp, addedRegex2));
                addedRegex2 = tmp;
            }
        }

        private static IEnumerable<Regex> GetInnerRegexList()
        {
            IList innerCacheList = GetInnerCacheList();

            foreach (object node in innerCacheList)
            {
                yield return GetRegex(node);
            }
        }

        private static IList GetInnerCacheList()
        {
            Type regexCacheType = Type.GetType("System.Text.RegularExpressions.RegexCache, System.Text.RegularExpressions")!;

            FieldInfo cacheListFieldInfo = regexCacheType.GetField("s_cacheList", BindingFlags.Static | BindingFlags.NonPublic)!;
            return (IList)cacheListFieldInfo.GetValue(null)!;
        }

        private static Regex GetRegex(object node)
        {
            Type nodeType = Type.GetType("System.Text.RegularExpressions.RegexCache+Node, System.Text.RegularExpressions")!;
            FieldInfo regexField = nodeType.GetField("Regex", BindingFlags.Public | BindingFlags.Instance)!;
            return (Regex)regexField.GetValue(node)!;
        }
```

### Expected behavior

Expect new regex `newlyAdded1` not be removed when add another new regex `newlyAdded2`, so that new data is treated as 'new' data and cache will be filled with new wave regex used recently.

### Actual behavior

Actually when cache is full, latest regex will override previous added 'new' regex, so that there will be only one new regex in cache, other recently used regex instances need to be re-compile again and again, which may impact expected effect of this cache and performance of related static methods of `Regex`. 

Following is output of repro code above:

```
When add addedPattern1:
addedRegex2 still in cache list ? - False
Regex instance 1 is reused ? - False
When add addedPattern2:
addedRegex1 still in cache list ? - False
Regex instance 2 is reused ? - False
When add addedPattern1:
addedRegex2 still in cache list ? - False
Regex instance 1 is reused ? - False
When add addedPattern2:
addedRegex1 still in cache list ? - False
Regex instance 2 is reused ? - False
```

### Regression?

Seems it always worked as this.

### Known Workarounds

_No response_

### Configuration

I test in .net 6, but should have same issue in other versions and environment.

### Other information

If the issue is valid, maybe the root cause should be cache implementation didn't assign latest virtual timestamp for newly added node. Currently, it is 0, even later the same regex is requested again, the 'fast path' field will still not update its timestamp because it thinks same continuous requests don't need to update its virtual timestamp again.

I raised [PR](https://github.com/dotnet/runtime/pull/105746) about it.

### PR DESCRIPTION
**Pull request by lateapexearlyspeed**
Fix #105748

### CODE REVIEW

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.Cache.cs [hunk 1]
```diff
@@ -254,7 +254,11 @@ private static void Add(Key key, Regex regex)
                 }
 
                 // Finally add the regex.
-                var node = new Node(key, regex);
+                var node = new Node(key, regex)
+                {
+                    LastAccessStamp = s_lastAccessed is { } lastAccessed ? Volatile.Read(ref lastAccessed.LastAccessStamp) + 1 : 1
```
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.Cache.cs:259]: If s_lastAccessed is null, so we need to set LastAccessedStamp? Why not just let it be 0 in that case?

Comment (lateapexearlyspeed) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.Cache.cs]: Yes, it can. Fixed, thanks.

**Review by steveharter — APPROVED** [review]
LGTM; thanks

**Comment by stephentoub** [comment]
Thanks!

**Pull request merged**